keps/sig-autoscaling/4943-vpa-heterogeneous-workloads/README.md

```diff
@@ -0,0 +1,911 @@
+-->
+
+One significant risk is scaling this feature. While VPA generally scales in a
+sublinear way because only a single recommendation is calculated, a VPA with a
+node scope will scale linerarly, requiring separate metrics queries for each
+node and separate recommendation calculations. Care will be taken to ensure
+this is handled efficiently.
+
+As of Kubernetes V1.32, [up to 5,000 nodes are
+supported](https://kubernetes.io/docs/setup/best-practices/cluster-large/) per
+cluster. There is no official recommendation on the maximum number of
+VerticalPodAutoscalers that are supported. However, this proposal does increase
+the load for VerticalPodAutoscalers that use the new node scope option.
+
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.  -->
+
+First, the implementation of this feature adds the `Scope` field to the
+`VerticalPodAutoscalerSpec` struct:
+
+```
+type VerticalPodAutoscalerSpec struct {
+	TargetRef *autoscaling.CrossVersionObjectReference `json:"targetRef" protobuf:"bytes,1,name=targetRef"`
+
+	UpdatePolicy *PodUpdatePolicy `json:"updatePolicy,omitempty" protobuf:"bytes,2,opt,name=updatePolicy"`
+
+	ResourcePolicy *PodResourcePolicy `json:"resourcePolicy,omitempty" protobuf:"bytes,3,opt,name=resourcePolicy"`
+
++   // Scope indicates at what level the recommendations should be calculated and applied.
++   // By default, the scope is all pods under the controller specified in `TargetRef`. Scope
++   // may also be set to "node" when the TargetRef is a DaemonSet, to apply recommendations
++   // independently to the pod scheduled to each node under the DaemonSet.
++   Scope string `json:"scope,omitzero" protobuf:"bytes,4,opt,name=scope"`
+}
+```
+
+An empty string in the scope field maintains existing functionality for the
+VerticalPodAutoscaler. The implied scope when the field is left empty is the
+entire TargetRef. Making scope a string type keeps the door open for different
+scopes to be supported in the future. For instance, although they are not
+supported in Kubernetes natively, several major cloud providers support the
+concept of NodePools. If a provider decides to support VPA at the NodePool
+level, they would be able to offer such a feature with minimal modification to
+the canoncial API.
+
+In addition, validation rules using OpenAPI schema and CEL will be added to the
+API to ensure that Scape is not set to an invalid value.
+
+When this feature is enabled, the internal representation of the
+VerticalPodAutoscaler will be divided into N objects, where N is the number of
+Nodes with a pod scheduled. Aside from this difference, most mechanics of
+VerticalPodAutoscaling will remain unchanged.
```

Comment:
What happens to the VerticalPoAutoscalerCheckPoint resource? I assume that still needs to be written to the cluster, and its lifecycle handled somehow.